### PR TITLE
[FIX] point_of_sale: fix rounding logic

### PR DESCRIPTION
**Pull request by kebeclibre**
Use case specifically for round globally, but the justifications hold anyway
Have a session with one tax, multiple partners, and returns of freight (negative amounts)

When validating a pos.session, the computation of amls was a bit out of logic,
Namely, we grouped tax lines by :
1. their tax id (correct)
2. their partner_id (not wrong but limited)
3. their balance > 0 (wrong)

The justification for point 2 and 3 is two fold:
- Ultimately, the order is level of granularity that really makes sense
It's the order that is paid and therefore that has both legal and business meaning
and on which the amounts are retrieved anyway
- On the order, we litterally make the sum of tax amounts of each line, aggregating
credit and debit under the same variable
It is only logic to do the same for grouping the taxes at validation time

OPW 1865700

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by kebeclibre** [comment]
@nim-odoo 
The second commit will certainly not be of your liking, but the logic is here, right ? ;)

**Comment by nim-odoo** [comment]
Let's discuss about it tomorrow. Although note that I think taxes are rounded all together on SO, and not on Invoices. So there might be issues elsewhere as well... To be checked in more details anyway.

File: addons/point_of_sale/models/pos_order.py [hunk 1]
```diff
@@ -380,7 +388,7 @@ def insert_data(data_type, values):
                         'credit': ((tax['amount'] > 0) and tax['amount']) or 0.0,
                         'debit': ((tax['amount'] < 0) and -tax['amount']) or 0.0,
                         'tax_line_id': tax['id'],
-                        'partner_id': partner_id
```
Comment (nim-odoo) [inline addons/point_of_sale/models/pos_order.py]: keep `partner_id`

File: addons/point_of_sale/models/pos_order.py [hunk 2]
```diff
@@ -321,6 +321,14 @@ def insert_data(data_type, values):
                         current_value['quantity'] = current_value.get('quantity', 0.0) + values.get('quantity', 0.0)
                         current_value['credit'] = current_value.get('credit', 0.0) + values.get('credit', 0.0)
                         current_value['debit'] = current_value.get('debit', 0.0) + values.get('debit', 0.0)
+                        if key[0] == 'tax':
+                            if current_value['debit'] - current_value['credit'] > 0:
```
Comment (nim-odoo) [inline addons/point_of_sale/models/pos_order.py]: use `float_compare`

File: addons/point_of_sale/models/pos_order.py [hunk 3]
```diff
@@ -194,10 +194,10 @@ def _get_account_move_line_group_data_type_key(self, data_type, values):
                     values['analytic_account_id'],
                     values['debit'] > 0)
         elif data_type == 'tax':
+            order_id = values.pop('order_id', False)
```
Comment (nim-odoo) [inline addons/point_of_sale/models/pos_order.py]: really necessary to remove?

File: addons/point_of_sale/models/pos_order.py [hunk 4]
```diff
@@ -194,10 +194,10 @@ def _get_account_move_line_group_data_type_key(self, data_type, values):
...
             return ('tax',
-                    values['partner_id'],
                     values['tax_line_id'],
-                    values['debit'] > 0)
+                    order_id)
```
Comment (nim-odoo) [inline addons/point_of_sale/models/pos_order.py]: only for `round_globally`

**Comment by kebeclibre** [comment]
@nim-odoo 
Changes made.
Yes it is "necessary" to pop order_id from the values, as it is not a field of account.move.line and will trigger warnings

File: addons/point_of_sale/models/pos_order.py [hunk 5]
```diff
@@ -380,8 +394,8 @@ def insert_data(data_type, values):
                         'credit': ((tax['amount'] > 0) and tax['amount']) or 0.0,
                         'debit': ((tax['amount'] < 0) and -tax['amount']) or 0.0,
                         'tax_line_id': tax['id'],
-                        'partner_id': partner_id
-                    })
+                        'partner_id': partner_id,
+                        'order_id': order.id})
```
Comment (nim-odoo) [inline addons/point_of_sale/models/pos_order.py]:
```python
    'order_id': order.id,
})
```

File: addons/point_of_sale/models/pos_order.py [hunk 6]
```diff
@@ -179,7 +179,7 @@ def _prepare_invoice(self):
         }
 
     @api.model
-    def _get_account_move_line_group_data_type_key(self, data_type, values):
+    def _get_account_move_line_group_data_type_key(self, data_type, values, options=False):
```
Comment (nim-odoo) [inline addons/point_of_sale/models/pos_order.py]: `options={}`

**Pull request merged**